### PR TITLE
Feature: Allow VPC Endpoints to be Created via vpc-endpoints Submodule

### PR DESCRIPTION
**Pull request by korenyoni**
## what
* Add `vpc-endpoints` submodule to this repo, which allows users to provision Interface and/or Gateway VPC Endpoints to the VPC they have created with the `terraform-aws-vpc` module.
* Add Terratest tests for testing the `vpc-endpoints` submodule
* Update automated testing via Terratest in this repo to use `us-east-2` region instead of `us-west-1`, as going forward CloudPosse repositories will be using `us-east-2`

## why
* Currently, there is no option within the VPC module to enable any VPC Endpoints. A simple switch for each type of endpoint is not possible, because there are dozens of possible endpoints —particularly Interface Endpoints— that can be created. Furthermore, VPC Endpoints may depend on resources that are themselves dependant on the `terraform-aws-vpc` module (i.e. subnet IDs), so an external module is required, but a standalone module is not necessarily justified.

## references
* It's useful to read the AWS Documentation to make a distinction between [Interface VPC Endpoints](https://docs.aws.amazon.com/vpc/latest/privatelink/vpce-interface.html) and [Gateway VPC Endpoints](https://docs.aws.amazon.com/vpc/latest/privatelink/vpce-gateway.html)

### CODE REVIEW

File: examples/vpc-endpoints/main.tf [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  vpc_id               = module.vpc.vpc_id
+  igw_id               = module.vpc.igw_id
+  cidr_block           = module.vpc.vpc_cidr_block
+  nat_gateway_enabled  = false
+  nat_instance_enabled = false
+
+  context = module.this.context
+}
+
+module "label" {
```
Comment (aknysh) [inline examples/vpc-endpoints/main.tf:67]: we don't need to create a separate label if it does not override anything from the existing `module.this.context` label.
Only if we want to override e.g. `name` or add `attributes`, we create a new label.
Please remove it and use 

module.this.is
module.this.tags

File: examples/vpc-endpoints/outputs.tf [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+output "vpc_cidr" {
+  value = module.vpc.vpc_cidr_block
+}
+
+output "vpc_id" {
+  value = module.vpc.vpc_id
+}
+
+output "gateway_vpc_endpoints" {
+  value = module.vpc-endpoints.gateway_vpc_endpoints
```
Comment (aknysh) [inline examples/vpc-endpoints/outputs.tf:18]:
```suggestion
  value = module.vpc_endpoints.gateway_vpc_endpoints
```

Let's be consistent and use `_` in all resource names

File: modules/vpc-endpoints/main.tf [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  service_type = "Gateway"
+}
+
+data "aws_vpc_endpoint_service" "interface_endpoint_service" {
+  for_each     = var.interface_vpc_endpoints
+  service      = var.interface_vpc_endpoints[each.key].name
+  service_type = "Interface"
+}
+
+module "label" {
```
Comment (aknysh) [inline modules/vpc-endpoints/main.tf:22]: same here, we don't need another label if we don't override anything in the existing `module.this` label

File: modules/vpc-endpoints/main.tf [hunk 4]
```diff
@@ -0,0 +1,53 @@
+locals {
+  enabled = var.context.enabled && (length(var.gateway_vpc_endpoints) > 0 || length(var.interface_vpc_endpoints) > 0)
```
Comment (aknysh) [inline modules/vpc-endpoints/main.tf:2]:
```suggestion
  enabled = module.this.enabled && (length(var.gateway_vpc_endpoints) > 0 || length(var.interface_vpc_endpoints) > 0)
```

**Review by aknysh — DISMISSED** [review]
LGTM, please see comments

**Comment by korenyoni** [comment]
Added a `do-not-merge` label because the `private_dns_enabled` attribute needs to be added.

File: README.yaml [hunk 5]
```diff
@@ -103,6 +103,46 @@ examples: |-
...
+          })
+        }
+      }
+      interface_vpc_endpoints = {
+        "ec2" = {
+          name               = "ec2"
+          security_group_ids = ["sg-12341234123412345"]
+          subnet_ids         = [module.dynamic_subnets.private_subnet_ids[0]]
+          policy             = null
+          private_dns_enabled = false
```
Comment (aknysh) [inline README.yaml:140]: `terraform fmt` the example :)

File: examples/vpc-endpoints/variables.tf [hunk 6]
```diff
@@ -0,0 +1,7 @@
+variable "region" {
+  type = string
```
Comment (aknysh) [inline examples/vpc-endpoints/variables.tf:2]: please add `description` to all variables and all check description for all outputs

File: examples/vpc-endpoints/outputs.tf [hunk 7]
```diff
@@ -0,0 +1,23 @@
+output "public_subnet_cidrs" {
+  value = module.subnets.public_subnet_cidrs
```
Comment (aknysh) [inline examples/vpc-endpoints/outputs.tf:2]: add `description` to all outputs

File: examples/vpc-endpoints/main.tf [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    "dynamodb" = {
+      name   = "dynamodb"
+      policy = null
+    }
+  }
+  interface_vpc_endpoints = {
+    "ec2" = {
+      name                = "ec2"
+      security_group_ids  = [aws_security_group.ec2_vpc_endpoint_sg.id]
+      subnet_ids          = [module.subnets.private_subnet_ids[0]]
```
Comment (aknysh) [inline examples/vpc-endpoints/main.tf:32]: why are we providing only one subnet even if the submodule takes a list?

File: modules/vpc-endpoints/main.tf [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+  enabled = module.this.enabled && (length(var.gateway_vpc_endpoints) > 0 || length(var.interface_vpc_endpoints) > 0)
+}
+
+data "aws_vpc" "vpc" {
+  count = local.enabled ? 1 : 0
+  id    = var.vpc_id
+}
+
+data "aws_vpc_endpoint_service" "gateway_endpoint_service" {
+  for_each     = var.gateway_vpc_endpoints
```
Comment (aknysh) [inline modules/vpc-endpoints/main.tf:11]: what if `module.this.enabled == false`, but `var.gateway_vpc_endpoints` or `var.interface_vpc_endpoints` is provided?
the data sources and resources with `for_each` will still be read and created.

You can use something like

```
for_each     = local.enabled ? var.gateway_vpc_endpoints : []
```

File: modules/vpc-endpoints/main.tf [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+  service      = var.gateway_vpc_endpoints[each.key].name
+  service_type = "Gateway"
+}
+
+data "aws_vpc_endpoint_service" "interface_endpoint_service" {
+  for_each     = var.interface_vpc_endpoints
```
Comment (aknysh) [inline modules/vpc-endpoints/main.tf:17]:
```suggestion
  for_each     = local.enabled ?  var.interface_vpc_endpoints : []
```

File: modules/vpc-endpoints/main.tf [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+  service      = var.interface_vpc_endpoints[each.key].name
+  service_type = "Interface"
+}
+
+resource "aws_vpc_endpoint" "gateway_endpoint" {
+  for_each          = data.aws_vpc_endpoint_service.gateway_endpoint_service
```
Comment (aknysh) [inline modules/vpc-endpoints/main.tf:23]: take into account `local.enabled`

File: modules/vpc-endpoints/main.tf [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+  vpc_endpoint_type = data.aws_vpc_endpoint_service.gateway_endpoint_service[each.key].service_type
+  vpc_id            = data.aws_vpc.vpc[0].id
+
+  tags = merge(
+    module.this.tags,
+  { Name = "${module.this.id}-${each.key}" })
+}
+
+resource "aws_vpc_endpoint" "interface_endpoint" {
+  for_each            = data.aws_vpc_endpoint_service.interface_endpoint_service
```
Comment (aknysh) [inline modules/vpc-endpoints/main.tf:35]: take into account `local.enabled`

File: modules/vpc-endpoints/main.tf [hunk 13]
```diff
@@ -0,0 +1,47 @@
+locals {
+  enabled = module.this.enabled && (length(var.gateway_vpc_endpoints) > 0 || length(var.interface_vpc_endpoints) > 0)
+}
+
+data "aws_vpc" "vpc" {
```
Comment (aknysh) [inline modules/vpc-endpoints/main.tf:5]: why do we need to read the VPC to get its ID if the VPC ID is already provided?

**Review by aknysh — CHANGES_REQUESTED** [review]
@korenyoni 
LGTM, a few more comment.
Alps please `terraform fmt` of all code, examples, and the code in README

File: examples/vpc-endpoints/main.tf (hunk 8 above)
Comment (korenyoni) [reply examples/vpc-endpoints/main.tf:32]: There is a limitation that there is a maximum of one subnet per AZ for which the Interface VPC Endpoint is deployed to. I've made this simple for example sake. Technically, the Interface VPC endpoint only needs to deploy an ENI to one AZ. Unless it is to be HA.

Comment (korenyoni) [reply examples/vpc-endpoints/main.tf]: Nevermind, I think this is something that affected the example during early development, and before switching the testing region to `us-east-2`. I've fixed this now.

File: modules/vpc-endpoints/main.tf [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+resource "aws_vpc_endpoint" "gateway_endpoint" {
+  for_each          = local.enabled ? data.aws_vpc_endpoint_service.gateway_endpoint_service : set()
+  service_name      = data.aws_vpc_endpoint_service.gateway_endpoint_service[each.key].service_name
+  policy            = var.gateway_vpc_endpoints[each.key].policy
+  vpc_endpoint_type = data.aws_vpc_endpoint_service.gateway_endpoint_service[each.key].service_type
+  vpc_id            = var.vpc_id
+
+  tags = merge(
+    module.this.tags,
+  { Name = "${module.this.id}-${each.key}" })
```
Comment (aknysh) [inline modules/vpc-endpoints/main.tf:26]: @korenyoni 

we usually don't manually override tags, we use a separate label module and add attributes to it

```
module "gateway_endpoint_label" {
  source  = "cloudposse/label/null"
  version = "0.24.1"

  for_each   = local.enabled ? data.aws_vpc_endpoint_service.gateway_endpoint_service : set()
  attributes = [each.key]

  context = module.this.context
}
```

and then

```
resource "aws_vpc_endpoint" "gateway_endpoint" {
  for_each          = local.enabled ? data.aws_vpc_endpoint_service.gateway_endpoint_service : set()
  service_name      = data.aws_vpc_endpoint_service.gateway_endpoint_service[each.key].service_name
  policy            = var.gateway_vpc_endpoints[each.key].policy
  vpc_endpoint_type = data.aws_vpc_endpoint_service.gateway_endpoint_service[each.key].service_type
  vpc_id            = var.vpc_id

  tags = module.gateway_endpoint_label.tags
```

@osterman  @Nuru  what do you think?

**Review by aknysh — COMMENTED** [review]
please see comments

**Review by aknysh — APPROVED** [review]

**Pull request merged**